tests/test_models.py

```diff
@@ -284,6 +291,185 @@ async def test_device_json(aresponses: ResponsesMockServer) -> None:
+
+        # Validate message start and end markers
+
+        socket_mock.send.side_effect = send_side_effect
+        socket_mock.recv.side_effect = lambda _: b"broken data"
+
+        with pytest.raises(OmnikInverterPacketInvalidError) as excinfo:
+            assert await client.inverter()
+
+        assert str(excinfo.value) == "Invalid start byte"
+
+        socket_mock.recv.side_effect = lambda _: bytearray([tcp.MESSAGE_START, 0])
+
+        with pytest.raises(OmnikInverterPacketInvalidError) as excinfo:
+            assert await client.inverter()
+
+        assert str(excinfo.value) == "Invalid end byte"
+
+        # Test checksum calculation through false checksum
+
+        test_message = b"foo"
+        false_checksum = 0xFE
+        checksum = sum(test_message) & 0xFF
+
+        socket_mock.recv.side_effect = lambda _: bytearray(
+            [tcp.MESSAGE_START] + list(test_message) + [false_checksum, tcp.MESSAGE_END]
+        )
+
+        with pytest.raises(OmnikInverterPacketInvalidError) as excinfo:
+            assert await client.inverter()
+
+        assert (
+            str(excinfo.value)
+            == f"Checksum mismatch (calculated `{checksum}` got `{false_checksum}`)"
+        )
+
+        # Test receiving valid data, but with mismatching serial number
+
+        def recv_side_effect(_max_bytes: int) -> bytes:
+            return load_fixture_bytes("tcp_reply.data")
+
+        socket_mock.recv.side_effect = recv_side_effect
+
+        with pytest.raises(OmnikInverterPacketInvalidError) as excinfo:
+            assert await client.inverter()
+
+        assert str(excinfo.value) == "Serial number mismatch in reply"
+
+    async def test_connection_broken(self) -> None:
+        """Test on connection broken after success - TCP source."""
+        serial_number = 1
+        socket_mock = asynctest.SocketMock()
+        socket_mock.type = socket.SOCK_STREAM
+
+        client = OmnikInverter(
+            host="example.com",
+            source_type="tcp",
+            serial_number=serial_number,
+            _socket_mock=socket_mock,
+        )
+
+        def send_side_effect(data: bytes) -> int:
+            assert data == tcp.create_information_request(serial_number)
+            asynctest.set_read_ready(socket_mock, self.loop)
+            return len(data)
+
+        socket_mock.send.side_effect = send_side_effect
+        socket_mock.recv.side_effect = Exception("Connection broken")
+
+        with pytest.raises(Exception) as excinfo:
+            assert await client.inverter()
+
+        assert str(excinfo.value) == "Connection broken"
+
+
+async def test_connection_failed() -> None:
+    """Test on failed connection attempt - TCP source."""
+    serial_number = 1
+
+    client = OmnikInverter(
+        host="example.com",
+        source_type="tcp",
+        serial_number=serial_number,
+        # Pass an invalid port number to simulate a failed connection attempt
+        tcp_port=123456,
+    )
```

Comment:
This test is terrible. I split `tcp_request` in two `try` blocks so that the `finally: writer.close()` of the second block doesn't run if `open_connection` - whose result is assigned to `writer` - fails (resulting in `` UnboundLocalError: local variable 'writer' referenced before assignment ``). Of course, `codecov` wants the new `except:` block to be tested :face_exhaling:).

It seems to be impossible to mock the `connect` function of the `SocketMock` since `open_connection` expects a `sock` that's already opened, so it won't try to call that method which we could otherwise use to "raise a failed connection attempt" here.

As an alternative I was hoping an obviously wrong argument - a port number above `65535` - would result in a validation failure directly inside `open_connection` but it turns out it "tries anyway" and only comes back with a connection error after a minute, locking up the test way too long :(